src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/CSharpVirtualDocumentFactory.cs

```diff
@@ -0,0 +1,88 @@
+
+using System;
+using System.Composition;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServerClient.Razor
+{
+    [Shared]
+    [Export(typeof(VirtualDocumentFactory))]
+    internal class CSharpVirtualDocumentFactory : VirtualDocumentFactory
+    {
+        // Internal for testing
+        internal const string CSharpLSPContentTypeName = "C#_LSP";
+        internal const string VirtualCSharpFileNameSuffix = "__virtual.cs";
+
+        private readonly IContentTypeRegistryService _contentTypeRegistry;
+        private readonly ITextBufferFactoryService _textBufferFactory;
+        private readonly FileUriProvider _fileUriProvider;
+        private IContentType _csharpLSPContentType;
+
+        [ImportingConstructor]
+        public CSharpVirtualDocumentFactory(
+            IContentTypeRegistryService contentTypeRegistry,
+            ITextBufferFactoryService textBufferFactory,
+            FileUriProvider fileUriProvider)
+        {
+            if (contentTypeRegistry is null)
+            {
+                throw new ArgumentNullException(nameof(contentTypeRegistry));
+            }
+
+            if (textBufferFactory is null)
+            {
+                throw new ArgumentNullException(nameof(textBufferFactory));
+            }
+
+            if (fileUriProvider is null)
+            {
+                throw new ArgumentNullException(nameof(fileUriProvider));
+            }
+
+            _contentTypeRegistry = contentTypeRegistry;
+            _textBufferFactory = textBufferFactory;
+            _fileUriProvider = fileUriProvider;
+        }
+
+        private IContentType CSharpLSPContentType
+        {
+            get
+            {
+                if (_csharpLSPContentType == null)
+                {
+                    _csharpLSPContentType = _contentTypeRegistry.GetContentType(CSharpLSPContentTypeName);
+                }
+
+                return _csharpLSPContentType;
+            }
+        }
+
+        public override bool TryCreateFor(ITextBuffer hostDocumentBuffer, out VirtualDocument virtualDocument)
+        {
+            if (hostDocumentBuffer is null)
+            {
+                throw new ArgumentNullException(nameof(hostDocumentBuffer));
+            }
+
+            if (!hostDocumentBuffer.ContentType.IsOfType(RazorLSPContentTypeDefinition.Name))
+            {
+                // Another content type we don't care about.
+                virtualDocument = null;
+                return false;
+            }
+
+            var hostDocumentUri = _fileUriProvider.GetOrCreate(hostDocumentBuffer);
+
+            // Index.cshtml => Index.cshtml__virtual.cs
+            var virtualCSharpFilePath = hostDocumentUri.AbsolutePath + VirtualCSharpFileNameSuffix;
+            var virtualCSharpUri = new Uri(virtualCSharpFilePath);
+
+            var csharpBuffer = _textBufferFactory.CreateTextBuffer(CSharpLSPContentType);
```

Comment:
@dibarbet @jasonmalinowski @tmat for now we're creating a C# buffer under the `C#_LSP` content type but we're not adding it to the VS workspace and I don't believe you guys automagically see this buffer and add it yourself.  

I know you all said that a new API would need to be made available for us to add content to your workspace. Is there something I can internally invoke for now to add this? Or would it be minor to expose/add the API?